hail/hail/ir-gen/src/Main.scala

```diff
@@ -0,0 +1,1108 @@
+          newChildren.slice(n, elt.nChildren(s"$self._${i + 1}")),
+        )
+      }.mkString("(", ", ", ")")
+    }
+  }
+
+  case class IR(
+    name: String,
+    attsAndChildren: Seq[NamedAttOrChildPack],
+    traits: Seq[Trait] = Seq.empty,
+    constraints: Seq[String] = Seq.empty,
+    extraMethods: Seq[String] = Seq.empty,
+    staticMethods: Seq[String] = Seq.empty,
+    docstring: String = "",
+    hasCompanionExtension: Boolean = false,
+  ) extends IR_Interface {
+    def withTraits(newTraits: Trait*): IR = copy(traits = traits ++ newTraits)
+    def withMethod(methodDef: String): IR = copy(extraMethods = extraMethods :+ methodDef)
+    def typed(typ: String): IR = withTraits(TypedIR(typ))
+    def withConstraint(c: String): IR = copy(constraints = constraints :+ c)
+
+    def withCompanionExtension: IR = copy(hasCompanionExtension = true)
+
+    def withClassExtension: IR = withTraits(Trait(s"${name}Ext"))
+
+    def withDocstring(docstring: String): IR = copy(docstring = docstring)
+
+    private def nChildren: NChildren = attsAndChildren.foldLeft(NChildren())(_ + _.nChildren)
+
+    private def children: String =
+      attsAndChildren.foldLeft[ChildrenSeq](ChildrenSeq.empty)(_ ++ _.childrenSeq).toString
+
+    private def childrenOffsets: Seq[NChildren] =
+      attsAndChildren.scanLeft(NChildren())(_ + _.nChildren)
+
+    private def copyMethod: String = {
+      val decl = s"override def copyWithNewChildren(newChildren: IndexedSeq[BaseIR]): $name = "
+      val assertion = s"assert(newChildren.length == $nChildren)"
+      val body = name + attsAndChildren.zipWithIndex.map { case (x, i) =>
+        try
+          x.pack.copyWithNewChildren(
+            x.name,
+            ChildrenSeqSlice(
+              ChildrenSeq.Dynamic("newChildren"),
+              nChildren,
+              childrenOffsets(i),
+              x.nChildren,
+            ),
+          )
+        catch {
+          case _: NotImplementedError =>
+            assert(false, name)
+        }
+      }.mkString("(", ", ", ")")
+      decl + "{\n    " + assertion + "\n    " + body + "}"
+    }
+
+    private def paramList = s"$name(${attsAndChildren.map(_.generateDeclaration).mkString(", ")})"
+
+    private def classDecl =
+      s"final case class $paramList extends IR" + traits.map(" with " + _.name).mkString
+
+    private def classBody = {
+      val childrenSeqDef = if (nChildren.hasStaticValue(0))
+        s"override def childrenSeq: IndexedSeq[BaseIR] = IndexedSeq.empty"
+      else
+        s"override lazy val childrenSeq: IndexedSeq[BaseIR] = $children"
+      val extraMethods =
+        this.extraMethods :+ childrenSeqDef :+ copyMethod
+      val constraints = this.constraints ++ attsAndChildren.flatMap(_.constraints)
+      if (constraints.nonEmpty || extraMethods.nonEmpty) {
+        (
+          " {" +
+            (if (constraints.nonEmpty)
+               constraints.map(c => s"  require($c)").mkString("\n", "\n", "\n")
+             else "")
```

Comment:
Do you need the extra newline characters? Couldn't you just fold over the `constraints`? 
```scala
constraints.foldLeft("") { (out, c) => out ++ s"  require($c)\n" }
constrains.view.map(c => s"  require($c)").mkString("\n")
```

Reply:
Sorry, I don't follow. Are both lines there part of the same suggestion?

One of the newlines there is just to put a blank line separating constraints from extra methods from standard generated methods. That's just a personal choice, and arguably we should configure the build system to always run scalafmt over the generated source anyways, but I haven't done that yet.